### ISSUE TITLE
BUG: Unable to load output from Kilosort4 into Phy

### ISSUE THREAD

**Issue by h21ak9** [MouseLand/Kilosort#604]
### Describe the issue:

Hello,

I am an engineer at Plexon. I used Kilosort4 to analyze a 20m37s recording made with a 256-channel high-density probe at a sampling frequency of 40,000 Hz. The analysis ran to completion, but I was unable to load the results in Phy. I did a little debugging and found the following:

•	templates.npy contains an array with dimensions 517x61x209 corresponding to 517 templates
•	spike_templates.npy contains 655 unique template id’s

So when Phy’s get_cluster_mean_waveforms() function calls get_template_id() for a spike_id larger than 516 (based on 0-based indexing), this breaks Phy.

Something is causing a mismatch between the data contained within templates.npy and spike_templates.npy. Would appreciate some suggestions as to how to proceed.

Thank you,
Nikhil Chandra

PS – I hope this is the correct forum to raise this sort of issue. Please do let me know.

### Reproduce the bug:

```python
I used the GUI to set up the analysis.
```

### Error message:

```shell
This is the error message that I get when I try to run Phy.

(phy) PS D:\data\File5\kilosort4_try1_failed> phy template-gui params.py
11:26:08.199 [E] __init__:62          An error has occurred (IndexError): index 519 is out of bounds for axis 0 with size 517
Traceback (most recent call last):
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\nikhil\miniconda3\envs\phy\Scripts\phy.exe\__main__.py", line 7, in <module>
    sys.exit(phycli())
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\click\decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phy\apps\__init__.py", line 159, in cli_template_gui
    template_gui(params_path, **kwargs)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phy\apps\template\gui.py", line 209, in template_gui
    model = load_model(params_path)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 1438, in load_model
    return TemplateModel(**get_template_params(params_path))
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 339, in __init__
    self._load_data()
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 419, in _load_data
    self.sparse_clusters = self.cluster_waveforms()
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 1324, in cluster_waveforms
    mean_waveform = self.get_cluster_mean_waveforms(clust, unwhiten=False)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 1224, in get_cluster_mean_waveforms
    for template_id in template_ids]
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 1224, in <listcomp>
    for template_id in template_ids]
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 961, in get_template
    unwhiten=unwhiten)
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\phylib\io\model.py", line 881, in _get_template_dense
    template_w = self.sparse_templates.data[template_id, ...]
  File "C:\Users\nikhil\miniconda3\envs\phy\lib\site-packages\numpy\core\memmap.py", line 334, in __getitem__
    res = super().__getitem__(index)
IndexError: index 519 is out of bounds for axis 0 with size 517
```

### Version information:

Windows 11
Python 3.9 for Kilosort4
Python 3.7 for Phy

### Context for the issue:

Users should have the ability to curate KS4's results in Phy.

**Comment by jacobpennington** [comment]
Hello Nikhil,
Please try the solutions I mentioned in issue #601 and let me know if that fixes the problem for you. The latest version of the code reverts some changes that were made to `spike_templates.npy`.

**Comment by h21ak9** [comment]
Hi Jacob,

Your fixed work. I uninstalled KS4 and reinstalled it from source. Now when I run it on my dataset, the results load properly in Phy. Thank you for your quick response!

However, there is another problem -- the results from KS4 are markedly different from KS3. Whereas KS3 returns templates that look somewhat reasonable (based on a visual assessment of their autocorrelograms), KS4 returns templates that look like noise / artifacts. I suspect something has gone wrong during the whitening preprocessing step. Attached are 2 images to illustrate what I mean. One shows what whitened data looks like in KS3. The other shows what whitened data looks like for the same file in KS4. Here are the parameters that I used in KS3:

Number of channels: 256
Sampling frequency (Hz): 40000
Time range (s): <left blank>
N blocks for registration: 5
Threshold: 9 9
Lambda (no longer very important): 10
AUC for splits (not used): 0.9

And here are the parameters from KS4:

number of channels: 256
sampling frequency: 40000
batch size: 60000
nblocks: 5
Th (universal): 9
Th (learned): 8
tmin: 0
tmax: inf

Here is the display for whitened data in KS3:
![ks3_whitened_display](https://github.com/MouseLand/Kilosort/assets/133174293/6ef6ec34-a7bb-40d2-9586-a8f703848ffe)

And here is the display in KS4:
![ks4_whitened_display](https://github.com/MouseLand/Kilosort/assets/133174293/ae3765b5-bd5b-40fb-b8dc-738f6b82a90c)

Your help is much appreciated. Thank you again!

**Comment by jacobpennington** [comment]
Hello Nikhil,
Usually something like the picture on the bottom happens if the number of channels is not inputted correctly, which will throw everything off. I'm not too familiar with the differences from KS3 so I'm not sure why using the same channel count for both wouldn't work, but @marius10p might have an idea.  For KS4 at least, "number of channels" needs to match the actual dimensions of the data in the binary file, even if some of the channels aren't ephys.

**Comment by h21ak9** [comment]
Hi Jacob,

The binary (RIL) file in question contains 12,668,679,424 data points which is exactly divisible by 256 -- I take this as confirmation that the file contains 256 channels' worth of data. This includes "disconnected" channels, for which the values filled in are zeros.

Here is a screenshot of my full KS4 GUI after loading everything. You can see that the number of channels I have entered is 256, but the diagonal-looking artifacts are still present.

![ks4_full_gui](https://github.com/MouseLand/Kilosort/assets/133174293/61d111fc-6d59-46b0-a41f-d9d9d18c6684)

Something is definitely not working properly -- why does it work for KS3 but not KS4?

Small side note -- I am unable to right click to enable/disable channels. Instead, I just get a menu that looks like this:

![ks4_menu](https://github.com/MouseLand/Kilosort/assets/133174293/c995ddff-e392-4627-a376-2e23bb314513)

**Comment by h21ak9** [comment]
I managed to fix the problem. When I loaded the channel map's .mat file, the "number of channels" field automatically updated to 256. Then when I pressed the giant LOAD button on top, I got the display with the diagonal artifacts. To correct this, I manually re-typed 256 in the "number of channels" box and hit LOAD again. Now the display looks like this:

![ks4_full_gui2](https://github.com/MouseLand/Kilosort/assets/133174293/e398371f-00f2-44c8-83f5-e7c7ef7f5549)

It looks like KS4 is stubbornly sticking with a value of 385 in this field behind the scenes, even when it loads the correct value of 256 from the channel map file.
